In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import scipy
import matplotlib.pyplot as plt

In [2]:
aq_df = pd.read_csv('Data.csv')
aq_df.head()

Timestamp                       station  PM2.5       PM10  \
0  2017-01-01    32Bungalows, Bhilai - CECB    NaN        NaN   
1  2017-01-01          AIIMS, Raipur - CECB    NaN        NaN   
2  2017-01-01  Adarsh Nagar, Jaipur - RSPCB    NaN        NaN   
3  2017-01-01    Airoli, Navi Mumbai - MPCB    NaN  54.743723   
4  2017-01-01    Airport Area, Indore - IMC    NaN        NaN   

                                             address         city   latitude  \
0  Bhilai RO Office, 5, 32 Bungalows, Bhilai, Dis...       Bhilai  21.194815   
1  All India Institute Of Medical Science, Gate N...       Raipur  21.258815   
2  Psychiatric Centre, Janta Colony, Govind Marg,...       Jaipur  26.902909   
3                                                NaN  Navi Mumbai  19.149400   
4                                                NaN          NaN        NaN   

   longitude         state  
0  81.314770  Chhattisgarh  
1  81.578979  Chhattisgarh  
2  75.836858     Rajasthan  
3  72.998600   Maharashtra  
4        NaN           NaN

#1.
NaN values are ignored automatically while using .mean()

In [29]:
grouped = aq_df.groupby('state', observed = True)['PM2.5']
display(grouped.mean())

state
Andhra Pradesh        35.049347
Arunachal Pradesh     22.674513
Assam                 57.962372
Bihar                 76.918759
Chandigarh            56.666803
Chhattisgarh          28.761571
Delhi                104.100918
Gujarat               52.462657
Haryana               72.371070
Himachal Pradesh      62.516878
Jammu and Kashmir     28.574696
Jharkhand             54.281545
Karnataka             28.894840
Kerala                29.404793
Madhya Pradesh        45.874375
Maharashtra           43.491415
Manipur               34.911856
Meghalaya             24.513390
Mizoram               11.784922
Nagaland              33.439550
Odisha                52.109504
Puducherry            22.854171
Punjab                52.243166
Rajasthan             54.230341
Sikkim                13.474286
Tamil Nadu            31.289620
Telangana             38.114469
Tripura               57.927642
Uttar Pradesh         70.893996
Uttarakhand           36.788767
West Bengal           54.611086
Na

In [30]:
display(grouped.mean().idxmax())

'Delhi'

Thus, Delhi has the highest average PM2.5 level across all years and all stations with 104.100918.

In [13]:
#2.
aq_df['Timestamp'] = pd.to_datetime(aq_df['Timestamp'])
display(aq_df.loc[aq_df['Timestamp'].dt.year == 2023])

Timestamp                           station      PM2.5        PM10  \
1229151 2023-01-01        32Bungalows, Bhilai - CECB        NaN         NaN   
1229152 2023-01-01              AIIMS, Raipur - CECB        NaN         NaN   
1229153 2023-01-01      Adarsh Nagar, Jaipur - RSPCB  52.502842   85.579053   
1229154 2023-01-01        Airoli, Navi Mumbai - MPCB        NaN         NaN   
1229155 2023-01-01        Airport Area, Indore - IMC        NaN         NaN   
...            ...                               ...        ...         ...   
1433911 2023-12-31              Worli, Mumbai - MPCB  47.100625  136.581481   
1433912 2023-12-31  Yamunapuram, Bulandshahr - UPPCB  67.694737  127.273684   
1433913 2023-12-31   Yerramukkapalli, Kadapa - APPCB        NaN         NaN   
1433914 2023-12-31  Zero Point GICI, Gangtok - SSPCB  10.938437   25.018125   
1433915 2023-12-31       Zoo Park, Hyderabad - TSPCB  60.600000   99.768421   

                                                   address         city  \
1229151  Bhilai RO Office, 5, 32 Bungalows, Bhilai, Dis...       Bhilai   
1229152  All India Institute Of Medical Science, Gate N...       Raipur   
1229153  Psychiatric Centre, Janta Colony, Govind Marg,...       Jaipur   
1229154                                                NaN  Navi Mumbai   
1229155                                                NaN          NaN   
...                                                    ...          ...   
1433911  89-B, NEERI Building, Dr. Annie Besant Road, W...       Mumbai   
1433912  F-5, Yamunapuram, Bulandshahr, Near MMR Mall (...  Bulandshahr   
1433913  Municipal Corporation Office Complex, Railway ...       Kadapa   
1433914  Near State PWD Office, Zero Point, Gangtok, Si...      Gangtok   
1433915                                                NaN    Hyderabad   

          latitude  longitude           state  
1229151  21.194815  81.314770    Chhattisgarh  
1229152  21.258815  81.578979    Chhattisgarh  
1229153  26.902909  75.836858       Rajasthan  
1229154  19.149400  72.998600     Maharashtra  
1229155        NaN        NaN             NaN  
...            ...        ...             ...  
1433911  18.993616  72.812811     Maharashtra  
1433912  28.406963  77.849831   Uttar Pradesh  
1433913  14.465052  78.824187  Andhra Pradesh  
1433914  27.338530  88.614099          Sikkim  
1433915  17.349694  78.451437       Telangana  

[204765 rows x 9 columns]

In [33]:
aq_df.loc[(aq_df['Timestamp'].dt.year == 2023) & (aq_df['PM2.5'] > 300)].groupby('state', observed = True)['Timestamp'].nunique()

state
Andhra Pradesh     2
Assam             21
Bihar             47
Delhi             49
Gujarat            4
Haryana           21
Jharkhand          1
Karnataka         14
Madhya Pradesh     8
Maharashtra       30
Odisha             1
Rajasthan         14
Tamil Nadu         3
Telangana          1
Tripura            1
Uttar Pradesh     24
Name: Timestamp, dtype: int64

In [34]:
aq_df.loc[(aq_df['Timestamp'].dt.year == 2023) & (aq_df['PM2.5'] > 300)].groupby('state', observed = True)['Timestamp'].nunique().idxmax()

'Delhi'

Thus, Delhi has had the most number of days with higher than 300 micrograms/m^3 of PM2.5 levels in 2023 with 49 such days.

In [26]:
#3.
aq_df.loc[(aq_df['Timestamp'].dt.year == 2023)].groupby('state', observed = True)['PM2.5'].std()

state
Andhra Pradesh       26.470621
Arunachal Pradesh    10.486218
Assam                61.501105
Bihar                63.734196
Chandigarh           37.715544
Chhattisgarh         20.076441
Delhi                82.797106
Gujarat              33.210780
Haryana              52.534649
Himachal Pradesh     45.573949
Jammu and Kashmir    17.571078
Jharkhand            40.211922
Karnataka            26.745988
Kerala               17.071595
Madhya Pradesh       40.026787
Maharashtra          39.960771
Manipur              31.033988
Meghalaya            22.994036
Mizoram               8.181937
Nagaland             20.281620
Odisha               36.130086
Puducherry           15.927181
Punjab               33.583028
Rajasthan            39.929352
Sikkim               19.351219
Tamil Nadu           22.017688
Telangana            18.842444
Tripura              46.884462
Uttar Pradesh        45.749190
Uttarakhand          20.194756
West Bengal          34.347047
Name: PM2.5, dtype: float64

In [28]:
aq_df.loc[(aq_df['Timestamp'].dt.year == 2023)].groupby('state', observed = True)['PM2.5'].std().idxmax()

'Delhi'

Yet again, we have Delhi at the top of the list. This time for most variability in PM2.5 levels across 2023.

In [24]:
#4.
aq_df.loc[(aq_df['Timestamp'].dt.year == 2020) | (aq_df['Timestamp'].dt.year == 2021)].groupby('state', observed = True)['PM2.5'].mean()

state
Andhra Pradesh        31.861788
Arunachal Pradesh     14.751325
Assam                 61.463611
Bihar                 75.120788
Chandigarh            39.586587
Chhattisgarh          24.863968
Delhi                100.613293
Gujarat               49.954478
Haryana               71.858425
Himachal Pradesh            NaN
Jammu and Kashmir     31.909474
Jharkhand             57.042476
Karnataka             27.604270
Kerala                26.407500
Madhya Pradesh        44.957182
Maharashtra           44.668592
Manipur                     NaN
Meghalaya             16.579606
Mizoram               14.323350
Nagaland              32.389944
Odisha                46.927677
Puducherry            20.198304
Punjab                49.561781
Rajasthan             56.182164
Sikkim                      NaN
Tamil Nadu            29.679225
Telangana             37.936372
Tripura               50.124667
Uttar Pradesh         87.086041
Uttarakhand                 NaN
West Bengal           55.807568
Na

In [25]:
aq_df.loc[(aq_df['Timestamp'].dt.year == 2020) | (aq_df['Timestamp'].dt.year == 2021)].groupby('state', observed = True)['PM2.5'].mean().idxmin()

'Mizoram'

Mizoram had the lowest average PM2.5 levels in the covid years.